### ISSUE TITLE
Stateful property + adaptivity (coarsening) + TET4/TRI3 = assertion

### ISSUE THREAD

**Issue by cpgr** [idaholab/moose#14434]
## Bug Description
I hit an assertion 
```
Assertion `props().contains(child_elem)' failed
Child element pointer is not in the MaterialProps data structure
at /Users/gre462/Documents/MOOSE/projects/moose/framework/src/materials/MaterialPropertyStorage.C, line 229
```
when trying to coarsen a previously refined mesh of either TET4 or TRI3 elements, see example input file below.

Unlike #14428, refining elements does work.

## Steps to Reproduce
The following input file (adapted from `test/tests/materials/stateful_prop/stateful_prop_adaptivity_test.i`) illustrates this bug. Initially, the mesh is uniformly refined (`uniform_refine = 2`), and then marked for coarsening in the `Adaptivity` block.

```
[Mesh]
  type = GeneratedMesh
  elem_type = TET4
  dim = 3
  nx = 2
  ny = 2
  nz = 2
  uniform_refine = 2
  # This option is necessary if you have uniform refinement + stateful material properties + adaptivity
  skip_partitioning = true
  # stateful material properties + adaptivity are not yet compatible
  # with distributed meshes
  parallel_type = replicated
[]

[Variables]
  [./u]
    order = FIRST
    family = LAGRANGE
  [../]
[]

[AuxVariables]
  [./prop1]
    order = CONSTANT
    family = MONOMIAL
  [../]
[]

[Kernels]
  [./heat]
    type = MatDiffusionTest
    variable = u
    prop_name = thermal_conductivity
    prop_state = old # Use the "Old" value to compute conductivity
  [../]
  [./ie]
    type = TimeDerivative
    variable = u
  [../]
[]

[AuxKernels]
  [./prop1_output]
    type = MaterialRealAux
    variable = prop1
    property = thermal_conductivity
  [../]
[]

[BCs]
  [./bottom]
    type = DirichletBC
    variable = u
    boundary = 1
    value = 0.0
  [../]
  [./top]
    type = DirichletBC
    variable = u
    boundary = 2
    value = 1.0
  [../]
[]

[Materials]
  [./stateful]
    type = StatefulTest
    prop_names = thermal_conductivity
    prop_values = 1.0
  [../]
[]

[Postprocessors]
  [./integral]
    type = ElementAverageValue
    variable = prop1
  [../]
[]

[Executioner]
  type = Transient

  solve_type = 'PJFNK'

  l_max_its = 10
  start_time = 0.0
  num_steps = 4
  dt = .1
  petsc_options_iname = '-pc_type -pc_hypre_type'
  petsc_options_value = 'hypre boomeramg'
[]

[Adaptivity]
  marker = uniform
  [./Markers]
    [./uniform]
      type = UniformMarker
      mark = coarsen
    [../]
  [../]
[]

[Outputs]
  exodus = true
  csv = true
[]

```
## Impact
Possibly affects some users?

**Comment by lindsayad** [comment]
I just tried this with #14167 and there are no assertions. @cpgr you can play around with that branch if you want

**Comment by cpgr** [comment]
Thanks @lindsayad. I won't bother looking into this one then.

**Comment by cpgr** [comment]
Just tried this again and it isn't asserting anymore. I assume that it was fixed by #14167 as above, so am closing this issue.
